server/photo.go

```diff
@@ -24,176 +27,279 @@ import (
+	ID              string     `storm:"id" json:"id"`
+	UploadedAt      *time.Time `storm:"index" json:"uploaded_at"`
+	TakenAt         *time.Time `storm:"index" json:"taken_at"`
+	Width           int        `storm:"index" json:"width"`
+	Height          int        `storm:"index" json:"height"`
+	Megapixels      float64    `storm:"index" json:"megapixels"`
+	Lat             float64    `json:"lat"`
+	Long            float64    `json:"long"`
+	CamSerial       string     `storm:"index" json:"cam_serial"`
+	CamMake         string     `storm:"index" json:"cam_make"`
+	CamModel        string     `storm:"index" json:"cam_model"`
+	Status          Status     `storm:"index" json:"status"`
+	StatusUpdatedAt *time.Time `storm:"index" json:"status_updated_at"`
 }
 
-func NewPhoto(id string, r image.Rectangle, x exif.Exif) *Photo {
+func (p *Photo) AddMetadata(r *image.Rectangle, x ExifData) {
 	taken, err := x.DateTime()
 	if err != nil {
-		log.Error("Unable to find time for ", id)
+		log.Error("unable to find time for ", p.ID)
+	} else {
+		p.TakenAt = &taken
 	}
 
 	lat, long, err := x.LatLong()
 	if err != nil {
-		log.Info("Unable to find gps for ", id)
+		log.Info("unable to find gps for ", p.ID)
+	} else {
+		p.Lat = lat
+		p.Long = long
 	}
 
 	cserial, err := x.Get(mknote.SerialNumber)
 	if err != nil {
-		log.Info("Unable to find serial for ", id)
+		log.Info("unable to find serial for ", p.ID)
+	} else {
+		p.CamSerial = strings.TrimSuffix(string(cserial.Val), "\u0000")
 	}
 
 	cmake, err := x.Get(exif.Make)
 	if err != nil {
-		log.Info("Unable to find make for ", id)
+		log.Info("unable to find make for ", p.ID)
+	} else {
+		p.CamMake = strings.TrimSuffix(string(cmake.Val), "\u0000")
 	}
 
 	cmodel, err := x.Get(exif.Model)
 	if err != nil {
-		log.Info("Unable to find model for ", id)
+		log.Info("unable to find model for ", p.ID)
+	} else {
+		p.CamModel = strings.TrimSuffix(string(cmodel.Val), "\u0000")
 	}
 
-	mp := toFixed(float64(r.Dx())*float64(r.Dy())/1000000.0, 1)
+	p.Width = r.Dx()
+	p.Height = r.Dy()
+	p.Megapixels = toFixed(float64(r.Dx())*float64(r.Dy())/1000000.0, 2)
+}
 
-	return &Photo{
-		ID:         id,
-		UploadedAt: time.Now(),
-		TakenAt:    taken,
-		Width:      r.Dx(),
-		Height:     r.Dy(),
-		Megapixels: mp,
-		Lat:        lat,
-		Long:       long,
-		CamSerial:  strings.TrimSuffix(string(cserial.Val), "\u0000"),
-		CamMake:    strings.TrimSuffix(string(cmake.Val), "\u0000"),
-		CamModel:   strings.TrimSuffix(string(cmodel.Val), "\u0000"),
-	}
+func (p *Photo) UpdateStatus(status Status) {
+	p.Status = status
+	t := time.Now()
+	p.StatusUpdatedAt = &t
 }
 
 func ProcessPhoto(input io.Reader, id string, photoPath string, thumbPath string) (*exif.Exif, *image.Rectangle, error) {
-	saveErr := make(chan error)
-	defer close(saveErr)
+	rect := make(chan *image.Rectangle, 1)
+	xif := make(chan *exif.Exif, 1)
+	errs := make(chan error, 3)
+	defer close(rect)
+	defer close(xif)
+	defer close(errs)
```

Comment:
These three deferred closes need to be changed. If the photo processing times out, the spawned goroutines will try to write to closed channels and panic. I think an acceptable fix is to delete these lines and never close the channels; the garbage collector will take care of them.